pyanaconda/ui/tui/spokes/software_selection.py

```diff
@@ -281,60 +281,32 @@ def ready(self):
                 not threadMgr.get(THREAD_SOFTWARE_WATCHER))
 
     def apply(self):
-        """ Apply our selections """
-        # no longer using values from kickstart
+        """Apply the changes."""
         self._kickstarted = False
-        self.data.packages.seen = True
-        # _apply depends on a value of _kickstarted
-        self._apply()
-
-    def _apply(self):
-        """ Private apply. """
-        self.environment = self._selected_environment
-        self.addons = self._addons_selection if self.environment is not None else set()
-
-        log.debug("Apply called old env %s, new env %s and addons %s",
-                  self._orig_env, self.environment, self.addons)
-
-        if self.environment is None:
-            return
-
-        changed = False
-
-        # Not a kickstart with packages, setup the selected environment and addons
-        if not self._kickstarted:
-
-            # Changed the environment or addons, clear and setup
-            if not self._orig_env \
-                    or self._orig_env != self.environment \
-                    or set(self._orig_addons) != set(self.addons):
-
-                log.debug("Setting new software selection old env %s, new env %s and addons %s",
-                          self._orig_env, self.environment, self.addons)
-
-                self.payload.data.packages.packageList = []
-                self.data.packages.groupList = []
-                self.payload.select_environment(self.environment)
 
-                environment_id = self._translate_env_name_to_id(self.environment)
-                available_addons = self._get_available_addons(environment_id)
+        # Clear packages data.
+        self._selection.packages = []
+        self._selection.excluded_packages = []
 
-                for addon_id in available_addons:
-                    if addon_id in self.addons:
-                        self.payload.select_group(addon_id)
+        # Clear groups data.
+        self._selection.excluded_groups = []
+        self._selection.groups_package_types = {}
 
-                changed = True
+        # Select valid groups.
+        # FIXME: Remove invalid groups from selected groups.
```

Comment:
Couldn't this be a big pain for users?

Reply:
GUI works this way. When you choose some groups and then change the environment, the invalid groups will break the transaction. You want to keep groups that are valid for the selected environment.